### PR TITLE
Civic Clerk site

### PR DESCRIPTION
**Pull request by fatima3558**
This PR adds a scraper for civic clerk sites.

## Testing
* run `python -m tests.civic_clerk_site` and confirm there are no assertion errors

### CODE REVIEW

**Comment by fatima3558** [comment]
@fgregg I spent a while looking at the lxml docs, trying to get my bearings. I was able to get data back from the request, but am having a hard time finding the tools I need in lxml to get the structured data I'm expecting back (there's a lot going on I don't understand). what are the high-level steps I should follow && what are the specific parts of the lmxl module I should explore more deeply to get the data I need?

**Comment by fatima3558** [comment]
I'm getting 500 errors on all the civic clerk instances I linked to in the issue thread. all code prior to the [`event_response` ](https://github.com/datamade/civic-scraper/pull/17/files#diff-52751ac16211312c8f7819983547729d64cfa1d3ed52c70c78718f34b997edb1R83) runs successfully

will start drafting the test for the last few minutes of the day, but will try other instances on Monday

File: civic_scraper/platforms/civic_clerk/site.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        link = event.xpath("./td[1]//a")[0]
+        href = link.attrib['href']
+        pattern = '.*?\((?P<id>.*?),.*'
+        match = re.match(pattern, href)
+        return 'civicclerk_{}_{}'.format(self.civicclerk_instance, match.group('id'))
+
+    def get_agenda_items(self, text, parser, session):
+        event_tree = etree.parse(StringIO(text), parser)
+        iframe = event_tree.xpath("//iframe")
+        if iframe:
```
Comment (fgregg) [inline civic_scraper/platforms/civic_clerk/site.py:55]: when would there not be an iframe? Add a comment here with an example url.

File: civic_scraper/platforms/civic_clerk/site.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            event_frame = iframe[0].attrib['src']
+            event_frame_url = 'https://chaffeecoco.civicclerk.com' + event_frame
+            frame_response = session.get(event_frame_url)
+            frame_tree = etree.parse(StringIO(frame_response.text), parser)
+
+            return frame_tree.xpath("//tr[./td[@class='dx-wrap dxtl dxtl__B0' and not(@colspan)]]")
+        return None
```
Comment (fgregg) [inline civic_scraper/platforms/civic_clerk/site.py:62]: don't need to explictly return None

File: civic_scraper/platforms/civic_clerk/site.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            str_datetime = event.xpath("./td[2]//text()")[0].strip()
+            meeting_datetime = datetime.strptime(str_datetime, '%m/%d/%Y %I:%M %p')
+            meeting_id = self.get_meeting_id(event)
+
+            event_url = f'{self.base_url}/Web/DocumentFrame.aspx?id={meeting_id}&mod=-1&player_tab=-2'
+            event_response = session.get(event_url)
+            agenda_items = self.get_agenda_items(event_response.text, parser, session) if event_response.status_code == 200 else None
+
+            # if event_response.status_code != 200:
+            #     print(event_response.status_code)
```
Comment (fgregg) [inline civic_scraper/platforms/civic_clerk/site.py:87]: remove commented out code

File: civic_scraper/platforms/civic_clerk/site.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+        for event in events:
+            committee_name = event.xpath("./td[1]//text()")[1].strip()
+            str_datetime = event.xpath("./td[2]//text()")[0].strip()
+            meeting_datetime = datetime.strptime(str_datetime, '%m/%d/%Y %I:%M %p')
+            meeting_id = self.get_meeting_id(event)
+
+            event_url = f'{self.base_url}/Web/DocumentFrame.aspx?id={meeting_id}&mod=-1&player_tab=-2'
+            event_response = session.get(event_url)
+            agenda_items = self.get_agenda_items(event_response.text, parser, session) if event_response.status_code == 200 else None
```
Comment (fgregg) [inline civic_scraper/platforms/civic_clerk/site.py:84]: don't love trinary operators, especially on long lines. would you please do it the other way? thanks!

Comment (fgregg) [reply civic_scraper/platforms/civic_clerk/site.py:84]: you could also handle this by having agenda_items return an empty list

File: civic_scraper/platforms/civic_clerk/site.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+            # if event_response.status_code != 200:
+            #     print(event_response.status_code)
+
+            if agenda_items:
```
Comment (fgregg) [inline civic_scraper/platforms/civic_clerk/site.py:89]: if agenda_items was an empty list wouldn't need this check.

File: civic_scraper/platforms/civic_clerk/site.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                for item in agenda_items:
+                    link_tr_text = item.xpath("./following-sibling::tr[1]")[0]
+                    link_tr = [(tr.attrib['href'], tr.xpath("./text()")[0]) for tr in link_tr_text.xpath(".//a") if tr.attrib['href'] != '#']
+                    if link_tr:
```
Comment (fgregg) [inline civic_scraper/platforms/civic_clerk/site.py:93]: don't need this check, since the right thing will happen if link_tr is an empty list

**Review by fgregg — CHANGES_REQUESTED** [review]
Really nicely organized methods!

File: civic_scraper/platforms/civic_clerk/site.py (hunk 1 above)
Comment (fatima3558) [reply civic_scraper/platforms/civic_clerk/site.py:55]: I added this when I was first encountering the 500 errors; my expectation is that if things run as expected, there will always be an iframe. I ran the tests now and I'm still getting those 500 errors. I'm going to address the comments I can, then investigate a bit. I might reach out to pair tomorrow if I can't figure out where the errors are coming from

**Comment by fgregg** [comment]
@fatima3558 would you give this a review?

**Comment by fatima3558** [comment]
@fgregg looks good to me & is working locally. The intricacies of the code that tackles pagination are going over my head a bit at the moment, would appreciate talking through it at some point.

File: civic_scraper/platforms/civic_clerk/site.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+                    ac.append(a)
+
+        if download:
+            asset_dir = Path(self.cache.path, 'assets')
+            asset_dir.mkdir(parents=True, exist_ok=True)
+            for asset in ac:
+                if asset.url:
+                    dir_str = str(asset_dir)
+                    asset.download(target_dir=dir_str, session=self.session)
+            # parse out info from this page
```
Comment (fatima3558) [inline civic_scraper/platforms/civic_clerk/site.py:208]: is this comment still necessary?

Comment (fgregg) [reply civic_scraper/platforms/civic_clerk/site.py]: handled! thanks!

**Pull request merged**